Storage/Stateflows.Storage.MongoDb/Stateflows/MongoDBStorage.cs

```diff
@@ -0,0 +1,124 @@
+﻿using MongoDB.Driver;
+using Newtonsoft.Json.Linq;
+using Stateflows.Common.Classes;
+using Stateflows.Common.Context;
+using Stateflows.Common.Interfaces;
+using Stateflows.Common.Utilities;
+using Stateflows.Storage.MongoDB.MongoDB.Entities;
+using Stateflows.Storage.MongoDB.Utils;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using TimeToken = Stateflows.Common.Classes.TimeToken;
+using TimeTokenEntity = Stateflows.Storage.MongoDB.MongoDB.Entities.TimeToken_v1;
+
+namespace Stateflows.Storage.MongoDB.Stateflows
+{
+    public class MongoDBStorage : IStateflowsStorage
+    {
+        private readonly IMongoDatabase _mongoDatabase;
+        public MongoDBStorage(IMongoClient mongoDBClient, string databaseName)
+        {
+            _mongoDatabase = mongoDBClient.GetDatabase(databaseName);
+        }
+
+        public async Task<StateflowsContext> Hydrate(BehaviorId id)
+        {
+            var context = await MongoDBSetExtension.FindOrCreate(_mongoDatabase, CollectionNames.Context, id);
+            StateflowsContext result = StateflowsJsonConverter.DeserializeObject<StateflowsContext>(context?.Data ?? string.Empty)
+                ?? new StateflowsContext() { Id = id };
+
+            return result;
+        }
+
+        public async Task Dehydrate(StateflowsContext context)
+        {
+            var contextEntity = await _mongoDatabase.FindOrCreate(CollectionNames.Context, context);
+            contextEntity.Data = StateflowsJsonConverter.SerializeObject(context);
+            await UpdateOrInsertContextData(contextEntity);
+        }
+
+        public async Task AddTimeTokens(TimeToken[] timeTokens)
+        {
+            if (!timeTokens.Any()) return;
+
+            var tokens = new Dictionary<TimeToken, TimeTokenEntity>();
+            var collection = _mongoDatabase.GetCollection<TimeTokenEntity>(CollectionNames.TimeToken);
+
+            foreach (var timeToken in timeTokens)
+            {
+                var timeTokenEntity = new TimeTokenEntity(
+                    StateflowsJsonConverter.SerializeObject(timeToken.TargetId.BehaviorClass),
+                    StateflowsJsonConverter.SerializeObject(timeToken)
+                );
+                tokens.Add(timeToken, timeTokenEntity);
+            }
+
+            await collection.InsertManyAsync(tokens.Values);
+            UpdateIdsForInsertedTimeTokens(tokens, timeTokens);
+        }
+
+        public async Task ClearTimeTokens(BehaviorId behaviorId, IEnumerable<string> ids)
```

Comment:
I don't know why, but it doesn't clear old time tokens - yes, I've tested it ;-)

Reply:
Could you check it now? :)

Reply:
Works ;-)